### PR TITLE
[Priest] Discipline bugfixes

### PR DESCRIPTION
**Pull request by taherbert**
This branches off of https://github.com/simulationcraft/simc/pull/7414 and includes:

* blaze of light refactor (auto apply not working)
* properly apply twist of fate to all specs
* fix warnings
* add shadow covenant's heal component (because it can trigger damage amp)
* remove unneeded trigger spell for borrowed time
* harsh discipline now uses trigger() instead of increment()
* power word: solace was missing an execute action
* override default expiation school from physical to shadow (in-game is shadow and benefits from buffs to shadow damage, spell data lists physical for some reason)
* expiation does not crit

### CODE REVIEW

File: engine/class_modules/priest/sc_priest.cpp [hunk 1]
```diff
@@ -523,20 +528,13 @@ struct smite_t final : public priest_spell_t
 
   smite_t( priest_t& p, util::string_view options_str )
     : priest_spell_t( "smite", p, p.find_class_spell( "Smite" ) ),
-      holy_fire_rank2( priest().find_rank_spell( "Holy Fire", "Rank 2" ) ),
       holy_word_chastise( priest().find_specialization_spell( 88625 ) ),
-      smite_rank2( priest().find_rank_spell( "Smite", "Rank 2" ) ),
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest.cpp:537]: remove these from the top of the struct as well please, lines 521 and 523

File: engine/class_modules/priest/sc_priest.cpp [hunk 2]
```diff
@@ -545,11 +543,22 @@ struct smite_t final : public priest_spell_t
     if ( priest().buffs.wrath_unleashed->check() )
     {
       d *= 1.0 + priest().buffs.wrath_unleashed->data().effectN( 1 ).percent();
+      sim->print_debug( "Smite damage modified by {} (new total: {}), from wrath unleashed",
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest.cpp]:
```suggestion
      sim->print_debug( "Smite damage modified by {} (new total: {}), from wrath_unleashed",
```

I would _highly_ suggest using `_` name format when you put spell names in debug. 99% of lines for a particular thing are formatted like this, so conforming to that makes it easier when searching. So in this case use `wrath_unleashed`

File: engine/class_modules/priest/sc_priest.cpp [hunk 3]
```diff
@@ -545,11 +543,22 @@ struct smite_t final : public priest_spell_t
...
+                        priest().buffs.wrath_unleashed->data().effectN( 1 ).percent(), d );
     }
     if ( priest().buffs.weal_and_woe->check() )
     {
       d *= 1.0 +
            ( priest().buffs.weal_and_woe->data().effectN( 1 ).percent() * priest().buffs.weal_and_woe->current_stack );
+      sim->print_debug(
+          "Smite damage modified by {} (new total: {}), from weal and woe",
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest.cpp]:
```suggestion
          "Smite damage modified by {} (new total: {}), from weal_and_woe",
```

File: engine/class_modules/priest/sc_priest.cpp [hunk 4]
```diff
@@ -545,11 +543,22 @@ struct smite_t final : public priest_spell_t
...
+          priest().buffs.weal_and_woe->data().effectN( 1 ).percent() * priest().buffs.weal_and_woe->current_stack, d );
+    }
+    if ( priest().talents.discipline.blaze_of_light.enabled() )
+    {
+      d *= 1.0 + ( priest().talents.discipline.blaze_of_light->effectN( 1 ).percent() );
+      sim->print_debug( "Smite damage modified by {} (new total: {}), from blaze of light",
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest.cpp]:
```suggestion
      sim->print_debug( "Smite damage modified by {} (new total: {}), from blaze_of_light",
```

File: engine/class_modules/priest/sc_priest_discipline.cpp [hunk 5]
```diff
@@ -67,6 +111,12 @@ struct penance_damage_t : public priest_spell_t
     {
       d *= 1.0 + priest().buffs.power_of_the_dark_side->data().effectN( 1 ).percent();
     }
+    if ( priest().talents.discipline.blaze_of_light.enabled() )
+    {
+      d *= 1.0 + ( priest().talents.discipline.blaze_of_light->effectN( 1 ).percent() );
+      sim->print_debug( "Penance damage modified by {} (new total: {}), from blaze of light",
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest_discipline.cpp]:
```suggestion
      sim->print_debug( "Penance damage modified by {} (new total: {}), from blaze_of_light",
```

File: engine/class_modules/priest/sc_priest_discipline.cpp [hunk 6]
```diff
@@ -408,7 +510,7 @@ void priest_t::create_buffs_discipline()
   buffs.harsh_discipline = make_buff( this, "harsh_discipline", talents.discipline.harsh_discipline )
                                ->set_max_stack( talents.discipline.harsh_discipline.enabled()
                                                     ? talents.discipline.harsh_discipline->effectN( 1 ).base_value()
-                                                    : 999 )
+                                                    : 0 )
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest_discipline.cpp]: you cannot set this to 0. If the talent is not found you make a buff with 0 stacks, which simc will complain about on every sim run (even shadow priest)

File: engine/class_modules/priest/sc_priest_discipline.cpp [hunk 7]
```diff
@@ -424,7 +526,7 @@ void priest_t::create_buffs_discipline()
 
   buffs.harsh_discipline_ready = make_buff( this, "harsh_discipline_ready", find_spell( 373183 ) );
 
-  buffs.borrowed_time = make_buff( this, "borrowed_time", find_spell( 390692 ) )->set_trigger_spell( find_spell( 17 ) );
+  buffs.borrowed_time = make_buff( this, "borrowed_time", find_spell( 390692 ) );
```
Comment (seanpeters86) [inline engine/class_modules/priest/sc_priest_discipline.cpp:530]: by removing this you will need to hook it into PW:S manually. Not sure if you've already done that

**Review by seanpeters86 — CHANGES_REQUESTED** [review]

File: engine/class_modules/priest/sc_priest_discipline.cpp (hunk 7 above)
Comment (taherbert) [reply engine/class_modules/priest/sc_priest_discipline.cpp:530]: Yes, PW:S triggers it manually.

File: engine/class_modules/priest/sc_priest_discipline.cpp (hunk 6 above)
Comment (taherbert) [reply engine/class_modules/priest/sc_priest_discipline.cpp]: Reverted to 999.

File: engine/class_modules/priest/sc_priest.cpp (hunk 1 above)
Comment (taherbert) [reply engine/class_modules/priest/sc_priest.cpp:537]: done

**Review by seanpeters86 — APPROVED** [review]

**Review by seanpeters86 — APPROVED** [review]

**Pull request merged**